### PR TITLE
LIPS GA

### PR DESCRIPTION
**Pull request by francastagna**
## PR Summary

Add LIPS (Linearly Independent Path) algorithm and tests

---

## LIPS Pseudocode (placeholder for breeder GA image)

image
---

## Main Changes

- EMConfig
  - Add `LIPS` in `EMConfig.Algorithm`.

- New algorithm
  - `core/src/main/kotlin/org/evomaster/core/search/algorithms/LIPSAlgorithm.kt`
    - Maintains a single active branch target (`currentTarget`) selected via `lastUncoveredBranchTargetId()`.
    - Per-target budgeting via `LipsBudget`; initializes a fair share per uncovered target; switches when covered or out of budget.
    - Next generation formation:
      - Elitism from current population (`formTheNextPopulation`).
      - Replace current population with `elites + offspring` for the next generation.

---

## Tests

- `core/src/test/kotlin/org/evomaster/core/search/algorithms/LIPSAlgorithmTest.kt`
  - Finds Optimum 
  - Edge cases
    - No crossover 
    - No mutation 
  - One-iteration properties

- `core/src/test/kotlin/org/evomaster/core/search/algorithms/LipsBudgetTest.kt`
  - Per-target fair-share budget
      - Actions: `computePerTargetBudget_Actions_FairShare` divides remaining evaluations evenly across uncovered targets; adapts as actions are spent.
      - Time: `computePerTargetBudget_Time_FairShare` divides remaining seconds evenly across uncovered targets.
  
    - Zero-uncovered handling
      - Actions: `computePerTargetBudget_Actions_ZeroUncovered_ReturnsRemaining` returns remaining evaluations to the current target when no uncovered targets.
      - Time: `computePerTargetBudget_Time_ZeroUncovered_UsesRemainingSeconds` returns remaining seconds to the current target when no uncovered targets.
  
    - Budget consumption and switching
      - Actions: `updateAndSwitchBudget` decrements by used actions; switches when budget reaches zero or when covered.
      - Time: `updateAndSwitchBudget_Time` decrements by elapsed seconds; switches when budget reaches zero or when covered.

---

### CODE REVIEW

File: core/src/main/kotlin/org/evomaster/core/search/service/Archive.kt [hunk 1]
```diff
@@ -616,6 +616,11 @@ class Archive<T> where T : Individual {
         return samplingCounter
     }
 
+    /**
+     * Expose the IdMapper used by the archive (read-only access pattern).
+     */
+    fun getIdMapper(): IdMapper = idMapper
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/search/service/Archive.kt]: why this? `IdMapper` is an injectable service singleton. you can inject it directly where needed

File: core/src/main/kotlin/org/evomaster/core/search/algorithms/LIPSAlgorithm.kt [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+    fun lastUncoveredBranchTargetId(): Int? {
+        val snapshot = archive.getSnapshotOfBestIndividuals()
+        if (snapshot.isEmpty()) return null
+
+        // Iterate targets by numeric id in descending order
+        val orderedIds = snapshot.keys.sortedDescending()
+
+        for (targetId in orderedIds) {
+            val description = archive.getIdMapper().getDescriptiveId(targetId)
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/search/algorithms/LIPSAlgorithm.kt]: why not just `@Inject` the `IdMapper` in this serviceclass?

**Review by arcuri82 — CHANGES_REQUESTED** [review]
